File: app/src/utils/similarity.ts
import type { NutrientFruit, NutrientKey } from '../types';
import { NUTRIENT_META } from './nutrition-meta';

interface SimilarFood {
  food: NutrientFruit;
  similarity: number;
}

const SIMILARITY_NUTRIENTS = NUTRIENT_META.filter(
  (m) => m.dailyValue !== null && m.key !== 'calories_kcal'
);

const MIN_SHARED_DIMENSIONS = 5;

function cosineSimilarity(a: NutrientFruit, b: NutrientFruit): number | null {
  let dot = 0;
  let magA = 0;
  let magB = 0;
  let shared = 0;

  for (const meta of SIMILARITY_NUTRIENTS) {
    const key = meta.key as NutrientKey;
    const va = a[key];
    const vb = b[key];
    if (va === null || vb === null) continue;

    const dv = meta.dailyValue!;
    const na = (va as number) / dv;
    const nb = (vb as number) / dv;

    dot += na * nb;
    magA += na * na;
    magB += nb * nb;
    shared++;
  }

  if (shared < MIN_SHARED_DIMENSIONS) return null;

  const denom = Math.sqrt(magA) * Math.sqrt(magB);
  if (denom === 0) return null;

  return dot / denom;
}

export function findSimilarFoods(
  target: NutrientFruit,
  allFoods: NutrientFruit[],
  count: number = 5
): SimilarFood[] {
  const results: SimilarFood[] = [];

  for (const food of allFoods) {
    if (food.name === target.name) continue;
    const similarity = cosineSimilarity(target, food);
    if (similarity === null) continue;
    results.push({ food, similarity });
  }

  results.sort((a, b) => b.similarity - a.similarity);
  return results.slice(0, count);
}

File: ios/Nutritionmaxx/Nutritionmaxx/Services/Similarity.swift
import Foundation

struct SimilarFood: Identifiable {
    let food: FoodItem
    let similarity: Double

    var id: String { food.name }
}

private let similarityNutrients = nutrientMetaList.filter {
    $0.dailyValue != nil && $0.key != .caloriesKcal
}

private let minSharedDimensions = 5

func findSimilarFoods(target: FoodItem, allFoods: [FoodItem], count: Int = 5) -> [SimilarFood] {
    var results: [SimilarFood] = []

    for food in allFoods {
        if food.name == target.name { continue }
        guard let similarity = cosineSimilarity(target, food) else { continue }
        results.append(SimilarFood(food: food, similarity: similarity))
    }

    results.sort { $0.similarity > $1.similarity }
    return Array(results.prefix(count))
}

private func cosineSimilarity(_ a: FoodItem, _ b: FoodItem) -> Double? {
    var dot = 0.0
    var magA = 0.0
    var magB = 0.0
    var shared = 0

    for meta in similarityNutrients {
        guard let va = a[meta.key], let vb = b[meta.key] else { continue }
        let dv = meta.dailyValue!
        let na = va / dv
        let nb = vb / dv
        dot += na * nb
        magA += na * na
        magB += nb * nb
        shared += 1
    }

    guard shared >= minSharedDimensions else { return nil }
    let denom = sqrt(magA) * sqrt(magB)
    guard denom > 0 else { return nil }
    return dot / denom
}
